font/src/ft/mod.rs

```diff
@@ -213,75 +222,56 @@ impl FreeTypeRasterizer {
-                self.get_specific_face(&desc, &style)
+                // If a name was specified, try and load specifically that font
+                pattern.add_style(style);
             },
         }
-    }
-
-    fn full_metrics(&self, key: FontKey) -> Result<FullMetrics, Error> {
-        let face = self.faces.get(&key).ok_or(Error::FontNotLoaded)?;
 
-        let size_metrics = face.ft_face.size_metrics().ok_or(Error::MissingSizeMetrics)?;
-
-        let width = match face.ft_face.load_char('0' as usize, face.load_flags) {
-            Ok(_) => face.ft_face.glyph().metrics().horiAdvance / 64,
-            Err(_) => size_metrics.max_advance / 64,
-        } as f64;
-
-        Ok(FullMetrics { size_metrics, cell_width: width })
-    }
-
-    fn get_matching_face(
-        &mut self,
-        desc: &FontDesc,
-        slant: Slant,
-        weight: Weight,
-    ) -> Result<FontKey, Error> {
-        let mut pattern = Pattern::new();
-        pattern.add_family(&desc.name);
-        pattern.set_weight(weight.into_fontconfig_type());
-        pattern.set_slant(slant.into_fontconfig_type());
-        pattern.add_pixelsize(self.pixel_size);
+        // Match pattern and get a font list, use first font as a "primary" one and the rest for a
+        // font fallback
+        let matched_fonts = fc::font_sort(&config, &mut pattern.clone())
+            .ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
+        let mut matched_fonts = matched_fonts.into_iter();
 
-        self.query_font(pattern, desc)
-    }
+        let primary_font =
+            matched_fonts.next().ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
 
-    fn get_specific_face(&mut self, desc: &FontDesc, style: &str) -> Result<FontKey, Error> {
-        let mut pattern = Pattern::new();
-        pattern.add_family(&desc.name);
-        pattern.add_style(style);
-        pattern.add_pixelsize(self.pixel_size);
+        // We should render patterns to get values like `pixelsizefixupfactor`
+        let primary_font = pattern.render_prepare(config, primary_font);
 
-        self.query_font(pattern, desc)
-    }
-
-    fn query_font(&mut self, pattern: Pattern, desc: &FontDesc) -> Result<FontKey, Error> {
-        let config = fc::Config::get_current();
-        let fonts = fc::font_sort(&config, &mut pattern.clone())
-            .ok_or_else(|| Error::MissingFont(desc.to_owned()))?;
+        // Compute ID for font identification, we're using hashes from requested pattern
+        // and a rendered one, and combining both with XOR to get a resulted ID. Some fonts
+        // always have the same size after pattern matching match, so if user requests pattern
+        // with size `X`, resulted font could always have `size = const`, which leads to
+        // constant result from e.g. `primary_font.hash()`, so we need to use requested
+        // `pixelsize` in the hash computation in addition to the hash of rendered pattern
```

Comment:
The location and length of this comment aren't quite right I'd say.  For the primary font this does not apply, right?

So I'd move it to the fallback fonts and shorten it a bit.

Maybe something like this should be sufficient:

```
// Hash fallback pattern together with request pattern to include requested font size in the hash
```

Reply:
It applies to all fonts, not only fallback ones, since you can request bitmap font.